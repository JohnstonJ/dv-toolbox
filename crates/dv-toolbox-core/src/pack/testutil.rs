use std::sync::LazyLock;

use display_error_chain::ErrorChainExt;
use googletest::prelude::*;
use num::rational::Ratio;

use super::*;
use crate::{
    file::{Info, UnvalidatedInfo},
    testutil::from_hex,
};

pub(crate) static NTSC: LazyLock<PackContext> = LazyLock::new(|| PackContext {
    file_info: UnvalidatedInfo::new(Info {
        file_size: 120_000,
        video_frame_rate: Ratio::<u32>::new(30_000, 1_001),
        video_duration: Ratio::<u128>::new(1_001, 30_000),
        audio_stereo_stream_count: 1,
        audio_sample_rate: Some(48_000),
    })
    .validate()
    .unwrap(),
});

pub(crate) static PAL: LazyLock<PackContext> = LazyLock::new(|| PackContext {
    file_info: UnvalidatedInfo::new(Info {
        file_size: 144_000,
        video_frame_rate: Ratio::<u32>::from(25),
        video_duration: Ratio::<u128>::new(1, 25),
        audio_stereo_stream_count: 1,
        audio_sample_rate: Some(48_000),
    })
    .validate()
    .unwrap(),
});

/// Shorthand function useful for constructing a validated pack from an unvalidated pack literal.
pub(crate) fn validated<T: PackData>(unvalidated_pack: T, ctx: PackContext) -> ValidPack<T> {
    Unvalidated::new(unvalidated_pack).validate_with(&ctx).unwrap().into()
}

#[derive(Debug)]
pub(crate) struct PackBinaryTestCase<'a> {
    pub(crate) input: &'a str,
    pub(crate) parsed: Option<Pack>,
    pub(crate) err: Option<&'a str>,
    pub(crate) output: Option<&'a str>,
    pub(crate) ctx: PackContext,
}

impl<'a> Default for PackBinaryTestCase<'a> {
    fn default() -> Self {
        Self { input: "", parsed: None, err: None, output: None, ctx: *NTSC }
    }
}

/// Test round trip of a pack: deserialize from binary, check the result, and then back to binary.
pub(crate) fn run_pack_binary_test_case(tc: &PackBinaryTestCase) {
    // Parse input hex to bytes
    let input = from_hex(tc.input);

    // Deserialize the pack and check for expected error and deserialized pack contents
    let (deserialized, err) = Pack::from_raw(&input, &tc.ctx);
    match tc.err {
        None => expect_that!(err, none()),
        Some(msg) => expect_that!(err.map(|e| e.chain().to_string()), some(eq(msg))),
    };
    let expected_pack = match tc.parsed {
        // Leaving tc.parsed as None is a shortcut that avoids retyping the Invalid pack.
        None => Pack::Invalid(
            Type::from(input[0]),
            ValidPack::<Unparsed>::try_from_raw(input[1..].try_into().unwrap(), &NTSC).unwrap(),
        ),
        Some(p) => p,
    };
    expect_that!(deserialized, eq(expected_pack));

    // Serialize the pack and check for expected output
    let serialized = deserialized.to_raw(&tc.ctx);
    // Leaving tc.output as None is a shortcut for saying that the input is expected
    let expected_output = tc.output.map_or_else(|| input, from_hex);
    expect_that!(serialized, eq(expected_output));
}
