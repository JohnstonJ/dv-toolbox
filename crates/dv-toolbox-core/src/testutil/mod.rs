use core::str;
use std::{
    collections::{HashMap, HashSet},
    path::PathBuf,
    sync::LazyLock,
};

use data_encoding::{Encoding, HEXUPPER_PERMISSIVE};
use derive_more::{Deref, From};
use googletest::prelude::*;
use regex::Regex;

/// Directory containing test-related data files.
pub(crate) fn test_resource(path: &str) -> PathBuf {
    [env!("CARGO_MANIFEST_DIR"), "resources/test", path].iter().collect()
}

static PADDED_HEX: LazyLock<Encoding> = LazyLock::new(|| {
    let mut spec = HEXUPPER_PERMISSIVE.specification();
    spec.ignore = String::from(" ");
    spec.encoding().unwrap()
});

/// Parse a string like "AA BB CC DD EE" into a byte array
pub(crate) fn from_hex<const N: usize>(value: &str) -> [u8; N] {
    PADDED_HEX
        .decode(value.as_bytes())
        .expect("value should be a valid hexadecimal string")
        .try_into()
        .unwrap()
}

/// Map of test case name to test case data.  See the [`test_case_map`] macro for an example of how
/// to use this.
#[derive(Debug, Deref, From)]
pub(crate) struct TestCaseMap<'a, T>(HashMap<&'a str, T>);

impl<'a, T> TestCaseMap<'a, T> {
    /// Return test case given the test function name generated by rstest
    ///
    /// The parameter needs to be the return value of the [`stdext::function_name`] macro.  The
    /// macro must be called from within a `#[case]` attribute.  Do not call it from within your
    /// test function itself.
    ///
    /// Example input: `my_crate_name::my_module::tests::test_frob::case_1_success_if_foo` which
    /// will return `success`.
    pub(crate) fn get_test_case(&self, test_function_name: &'a str) -> &T {
        let case_name = &GET_TEST_CASE_RE.captures(test_function_name).unwrap()[1];
        assert!(self.contains_key(case_name), "Test case {case_name} is not in test table");
        &self[case_name]
    }
}

/// A lazy-initialized map of test cases.  See the [`test_case_map`] macro for an example of how to
/// use this.
pub(crate) type LazyTestCases<'a, T> = LazyLock<TestCaseMap<'a, T>>;

/// Lazy-initializes a map of [`rstest`] test cases for a given function.  Intended for use as a
/// static variable.
///
/// The advantage of using this over passing structures directly to `#[case]` is that the structures
/// will be auto-formatted with rustfmt.
///
/// # Examples
///
/// ```ignore
/// use rstest::rstest;
/// use stdext::function_name;
///
/// #[derive(Debug)]
/// struct FunctionATestCase {
///     foo: i32,
///     bar: i32,
/// }
///
/// static FUNCTION_A_TEST_CASES: LazyTestCases<FunctionATestCase> = test_case_map!(
///     "first_case",
///     FunctionATestCase { foo: 1, bar: 1 },
///     "second_case",
///     FunctionATestCase { foo: 2, bar: 2 }
/// );
///
/// #[rstest]
/// #[case::first_case(function_name!())]
/// #[case::second_case(function_name!())]
/// fn test_function_a(#[case] test_function_name: &str) {
///     let tc = FUNCTION_A_TEST_CASES.get_test_case(test_function_name);
///
///     assert!(tc.foo > 0);
/// }
///
/// // If you add an entry to FUNCTION_A_TEST_CASES but forget to add a #[case] for it, this will
/// // catch the mistake.
/// test_all_test_cases_ran!(
///     ("test_function_a", &FUNCTION_A_TEST_CASES)
/// );
/// ```
macro_rules! test_case_map {
    ($($test_case_name:expr, $test_case_data:expr),*) => {
        ::std::sync::LazyLock::new(|| ::std::collections::HashMap::<&str, _>::from([
            $(($test_case_name, $test_case_data)),*
        ]).into())
    };
}

pub(crate) use test_case_map;

/// Asserts that all [`rstest`] test cases in the given tables of test cases are actual test cases
/// in this test binary.
///
/// If you are using the [`test_case_map`] macro and [`TestCaseMap`] type, then you should also
/// call this macro to make sure that every entry in a [`TestCaseMap`] has a corresponding `case`
/// statement in [`rstest::rstest`].
///
/// # Examples
///
/// ```ignore
/// test_all_test_cases_ran!(
///     ("test_function_a", &FUNCTION_A_TEST_CASES),
///     ("test_function_b", &FUNCTION_B_TEST_CASES),
/// );
/// ```
macro_rules! test_all_test_cases_ran {
    ($($test_function_test_cases:expr),*) => {
        #[test]
        fn test_all_test_cases_ran() {
            let mut _expected_test_cases = ::std::collections::HashMap::<&str, Vec<&str>>::new();
            $({
                let (test_function_name, test_case_map) = $test_function_test_cases;
                _expected_test_cases
                    .insert(test_function_name, test_case_map.keys().map(|k| *k).collect());
            })*
            $crate::testutil::assert_all_test_cases_ran(&_expected_test_cases, module_path!());
        }
    };
}

pub(crate) use test_all_test_cases_ran;

static GET_TEST_CASE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r".+::case_\d*_([^:]+)$").unwrap());

static GET_CRATE_NAME_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^[^:]+").unwrap());

static PARSE_TEST_LIST_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^(.+)::case_\d*_(.+): test$").unwrap());

/// Asserts that all [`rstest`] test cases in the given tables of test cases are actual test cases
/// in this test binary.
///
/// The function is not generally to be used directly.  Use the [`test_all_test_cases_ran`] macro
/// instead.
pub(crate) fn assert_all_test_cases_ran(
    expected_test_cases: &HashMap<&str, Vec<&str>>,
    test_module_path: &str,
) {
    // final result for each set are paths like:
    // my_crate_name::my_module::tests::test_frob::success_if_foo

    // test cases built from the user's tast table HashMap
    let expected_test_cases = HashSet::<String>::from_iter(expected_test_cases.iter().flat_map(
        |(test_function_name, test_cases_for_function)| {
            test_cases_for_function.iter().map(move |test_case_name| {
                format!("{test_module_path}::{test_function_name}::{test_case_name}")
            })
        },
    ));

    let crate_name = GET_CRATE_NAME_RE.find(test_module_path).unwrap().as_str();

    // run the test executable to list all tests
    // the output of this command looks like:
    // my_module::tests::test_frob::case_02_success_if_foo: test
    let list_all_tests_output = std::process::Command::new(std::env::current_exe().unwrap())
        .args(["--list", "--format", "terse"])
        .output()
        .unwrap();
    assert_that!(
        list_all_tests_output.status.success(),
        eq(true),
        "Failed to run test EXE to list tests"
    );
    let actual_test_cases = HashSet::<String>::from_iter(
        str::from_utf8(&list_all_tests_output.stdout).unwrap().lines().flat_map(
            |test_output_line| {
                let parsed = PARSE_TEST_LIST_RE.captures(test_output_line)?;
                Some(format!("{crate_name}::{}::{}", &parsed[1], &parsed[2]))
            },
        ),
    );

    let unrun_test_cases: Vec<_> = expected_test_cases.difference(&actual_test_cases).collect();
    assert_that!(
        unrun_test_cases,
        empty(),
        "These test cases are not referenced by the test function with a #[case] attribute."
    );
}

mod tests {
    use googletest::expect_that;

    use super::*;

    #[googletest::test]
    fn test_from_hex() {
        expect_that!(from_hex("AA BB CC"), eq([0xAA, 0xBB, 0xCC]));
    }
}
