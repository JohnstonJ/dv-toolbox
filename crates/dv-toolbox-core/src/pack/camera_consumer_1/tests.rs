use arbitrary_int::{u2, u4};
use insta::{assert_debug_snapshot, with_settings};
use itertools::Itertools;
use rstest::rstest;
use rust_decimal_macros::dec;
use stdext::function_name;
use testutil::*;

use super::super::*;
use crate::{pack::testutil::PackBinaryTestCase, testutil::*};

test_all_test_cases_ran!(
    ("test_camera_consumer_1_binary", &CAMERA_CONSUMER_1_BINARY_TEST_CASES,),
    ("test_camera_consumer_1_validation", &CAMERA_CONSUMER_1_VALIDATION_TEST_CASES)
);

// ==================== BINARY SERIALIZATION TESTING ====================
// Tests to/from actual/raw DV pack data.

static CAMERA_CONSUMER_1_BINARY_TEST_CASES: LazyTestCases<PackBinaryTestCase> = test_case_map!(
    //
    // ===== SOME BASIC SUCCESS CASES =====
    // These are from real tape transfers
    //
    // basic success case: from my Sony DCR-TRV460
    "basic_success_dcr_trv460",
    PackBinaryTestCase {
        input: "70 C5 07 1F FF",
        parsed: Some(Pack::CameraConsumer1(validated(
            CameraConsumer1 {
                auto_exposure_mode: Some(AutoExposureMode::FullAutomatic),
                iris: Some(dec!(1.5)),
                auto_gain_control: Some(u4::new(7)),
                white_balance_mode: Some(WhiteBalanceMode::Automatic),
                white_balance: None,
                focus_mode: FocusMode::Manual,
                focus_position: None,
                reserved: u2::new(0x3),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // basic success case: from a tape originating from some old Digital8 camcorder 20 years ago
    "basic_success_old_camcorder",
    PackBinaryTestCase {
        input: "70 DE 01 1F FF",
        parsed: Some(Pack::CameraConsumer1(validated(
            CameraConsumer1 {
                auto_exposure_mode: Some(AutoExposureMode::FullAutomatic),
                iris: Some(dec!(13.5)),
                auto_gain_control: Some(u4::new(1)),
                white_balance_mode: Some(WhiteBalanceMode::Automatic),
                white_balance: None,
                focus_mode: FocusMode::Manual,
                focus_position: None,
                reserved: u2::new(0x3),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // ===== ADDITIONAL CONTRIVED/SYNTHETIC TEST CASES =====
    //
    // everything's kind of in the middle here
    "all_values_in_range",
    PackBinaryTestCase {
        input: "70 E5 37 44 56",
        parsed: Some(Pack::CameraConsumer1(validated(
            CameraConsumer1 {
                auto_exposure_mode: Some(AutoExposureMode::IrisPriority),
                iris: Some(dec!(24.7)),
                auto_gain_control: Some(u4::new(7)),
                white_balance_mode: Some(WhiteBalanceMode::OnePush),
                white_balance: Some(WhiteBalance::Sunlight),
                focus_mode: FocusMode::Automatic,
                focus_position: Some(2100),
                reserved: u2::new(0x3),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // maximum values that are calculated (especially iris/focus position)
    "max_calculated_values",
    PackBinaryTestCase {
        input: "70 FC 3E 44 7E",
        parsed: Some(Pack::CameraConsumer1(validated(
            CameraConsumer1 {
                auto_exposure_mode: Some(AutoExposureMode::IrisPriority),
                iris: Some(dec!(181.0)),
                auto_gain_control: Some(u4::new(14)),
                white_balance_mode: Some(WhiteBalanceMode::OnePush),
                white_balance: Some(WhiteBalance::Sunlight),
                focus_mode: FocusMode::Automatic,
                focus_position: Some(3100),
                reserved: u2::new(0x3),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // wide open iris
    "wide_open_iris",
    PackBinaryTestCase {
        input: "70 FD 37 44 7E",
        parsed: Some(Pack::CameraConsumer1(validated(
            CameraConsumer1 {
                auto_exposure_mode: Some(AutoExposureMode::IrisPriority),
                iris: Some(dec!(0.0)),
                auto_gain_control: Some(u4::new(7)),
                white_balance_mode: Some(WhiteBalanceMode::OnePush),
                white_balance: Some(WhiteBalance::Sunlight),
                focus_mode: FocusMode::Automatic,
                focus_position: Some(3100),
                reserved: u2::new(0x3),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // closed iris
    "closed_iris",
    PackBinaryTestCase {
        input: "70 FE 37 44 7E",
        parsed: Some(Pack::CameraConsumer1(validated(
            CameraConsumer1 {
                auto_exposure_mode: Some(AutoExposureMode::IrisPriority),
                iris: Some(dec!(999.9)),
                auto_gain_control: Some(u4::new(7)),
                white_balance_mode: Some(WhiteBalanceMode::OnePush),
                white_balance: Some(WhiteBalance::Sunlight),
                focus_mode: FocusMode::Automatic,
                focus_position: Some(3100),
                reserved: u2::new(0x3),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // all bits set
    "all_bits_set",
    PackBinaryTestCase {
        input: "70 FF FF FF FF",
        parsed: Some(Pack::CameraConsumer1(validated(
            CameraConsumer1 {
                auto_exposure_mode: None,
                iris: None,
                auto_gain_control: None,
                white_balance_mode: None,
                white_balance: None,
                focus_mode: FocusMode::Manual,
                focus_position: None,
                reserved: u2::new(0x3),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    },
    //
    // all bits clear
    "all_bits_clear",
    PackBinaryTestCase {
        input: "70 00 00 00 00",
        parsed: Some(Pack::CameraConsumer1(validated(
            CameraConsumer1 {
                auto_exposure_mode: Some(AutoExposureMode::FullAutomatic),
                iris: Some(dec!(1.0)),
                auto_gain_control: Some(u4::new(0)),
                white_balance_mode: Some(WhiteBalanceMode::Automatic),
                white_balance: Some(WhiteBalance::Candle),
                focus_mode: FocusMode::Automatic,
                focus_position: Some(0),
                reserved: u2::new(0x0),
            },
            *NTSC
        ))),
        ctx: *NTSC,
        ..Default::default()
    }
);

#[googletest::test]
#[rstest]
#[case::basic_success_dcr_trv460(function_name!())]
#[case::basic_success_old_camcorder(function_name!())]
#[case::all_values_in_range(function_name!())]
#[case::max_calculated_values(function_name!())]
#[case::wide_open_iris(function_name!())]
#[case::closed_iris(function_name!())]
#[case::all_bits_set(function_name!())]
#[case::all_bits_clear(function_name!())]
fn test_camera_consumer_1_binary(#[case] test_function_name: &str) {
    let tc = CAMERA_CONSUMER_1_BINARY_TEST_CASES.get_test_case(test_function_name);
    run_pack_binary_test_case(tc);
}

// ==================== VALIDATION TESTING ====================
// Tests on validation code that was not already tested as part of binary serialization.

static CAMERA_CONSUMER_1_VALIDATION_TEST_CASES: LazyTestCases<
    ValidateFailureTestCase<CameraConsumer1, PackContext>,
> = test_case_map!(
    //
    // unsupported iris value
    "invalid_iris",
    ValidateFailureTestCase {
        value: CameraConsumer1 {
            auto_exposure_mode: Some(AutoExposureMode::FullAutomatic),
            iris: Some(dec!(50.2)),
            auto_gain_control: Some(u4::new(7)),
            white_balance_mode: Some(WhiteBalanceMode::Automatic),
            white_balance: None,
            focus_mode: FocusMode::Manual,
            focus_position: None,
            reserved: u2::new(0x3),
        },
        err: "iris: iris value 50.2 not supported: only iris values returned by the \
            valid_iris_values function are supported\n",
        ctx: *NTSC
    },
    //
    // unsupported iris value
    "invalid_auto_gain_control",
    ValidateFailureTestCase {
        value: CameraConsumer1 {
            auto_exposure_mode: Some(AutoExposureMode::FullAutomatic),
            iris: Some(dec!(1.5)),
            auto_gain_control: Some(u4::new(15)),
            white_balance_mode: Some(WhiteBalanceMode::Automatic),
            white_balance: None,
            focus_mode: FocusMode::Manual,
            focus_position: None,
            reserved: u2::new(0x3),
        },
        err: "auto_gain_control: instead of specifying Some(0xF), use None to \
            indicate no information\n",
        ctx: *NTSC
    },
    //
    // unsupported focus position
    "invalid_focus_position",
    ValidateFailureTestCase {
        value: CameraConsumer1 {
            auto_exposure_mode: Some(AutoExposureMode::FullAutomatic),
            iris: Some(dec!(1.5)),
            auto_gain_control: Some(u4::new(7)),
            white_balance_mode: Some(WhiteBalanceMode::Automatic),
            white_balance: None,
            focus_mode: FocusMode::Manual,
            focus_position: Some(1234),
            reserved: u2::new(0x3),
        },
        err: "focus_position: focus position 1234 not supported: only focus positions \
            returned by the valid_focus_positions function are supported\n",
        ctx: *NTSC
    }
);

#[googletest::test]
#[rstest]
#[case::invalid_iris(function_name!())]
#[case::invalid_auto_gain_control(function_name!())]
#[case::invalid_focus_position(function_name!())]
fn test_camera_consumer_1_validation(#[case] test_function_name: &str) {
    let tc = CAMERA_CONSUMER_1_VALIDATION_TEST_CASES.get_test_case(test_function_name);
    run_validate_failure_test_case(tc);
}

// ==================== VALID VALUES SNAPSHOT TESTING ====================
// Tests that the valid values match a snapshot, which should be compared vs the specified values in
// the IEC specification.

#[googletest::test]
fn test_valid_iris_values() {
    let displayed_iris_values: Vec<String> =
        CameraConsumer1::valid_iris_values().iter().sorted().map(|i| format!("{}", i)).collect();
    with_settings!({prepend_module_to_snapshot => false}, {
        assert_debug_snapshot!(displayed_iris_values);
    });
}

#[googletest::test]
fn test_valid_focus_positions() {
    let displayed_focus_positions: Vec<String> = CameraConsumer1::valid_focus_positions()
        .iter()
        .sorted()
        .map(|i| format!("{}", i))
        .collect();
    with_settings!({prepend_module_to_snapshot => false}, {
        assert_debug_snapshot!(displayed_focus_positions);
    });
}
