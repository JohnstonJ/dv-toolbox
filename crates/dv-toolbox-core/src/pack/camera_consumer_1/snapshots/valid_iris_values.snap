---
source: crates/dv-toolbox-core/src/pack/camera_consumer_1/tests.rs
expression: displayed_iris_values
---
[
    "0.0",
    "1",
    "1.1",
    "1.2",
    "1.3",
    "1.4",
    "1.5",
    "1.7",
    "1.8",
    "2",
    "2.2",
    "2.4",
    "2.6",
    "2.8",
    "3.1",
    "3.4",
    "3.7",
    "4",
    "4.4",
    "4.8",
    "5.2",
    "5.7",
    "6.2",
    "6.7",
    "7.3",
    "8",
    "8.7",
    "9.5",
    "10.4",
    "11.3",
    "12.3",
    "13.5",
    "14.7",
    "16",
    "17.4",
    "19.0",
    "20.7",
    "22.6",
    "24.7",
    "26.9",
    "29.3",
    "32",
    "34.9",
    "38.1",
    "41.5",
    "45.3",
    "49.4",
    "53.8",
    "58.7",
    "64",
    "69.8",
    "76.1",
    "83.0",
    "90.5",
    "98.7",
    "107.6",
    "117.4",
    "128",
    "139.6",
    "152.2",
    "166.0",
    "181.0",
    "999.9",
]
